File: examples/client-example.js
#!/usr/bin/env node
import { McpClient } from '@modelcontextprotocol/sdk/client/index.js';
import { ChildProcessClientTransport } from '@modelcontextprotocol/sdk/client/child-process.js';

/**
 * Example client for the Mac Shell MCP Server
 * This demonstrates how to connect to the server and use its tools
 */
async function main() {
  // Create a client transport that connects to the server
  const transport = new ChildProcessClientTransport({
    command: 'node',
    args: ['../build/index.js'],
  });

  // Create the MCP client
  const client = new McpClient();
  
  try {
    // Connect to the server
    await client.connect(transport);
    console.log('Connected to Mac Shell MCP Server');

    // List available tools
    const tools = await client.listTools();
    console.log('Available tools:');
    tools.forEach(tool => {
      console.log(`- ${tool.name}: ${tool.description}`);
    });

    // Get the whitelist
    console.log('\nWhitelisted commands:');
    const whitelist = await client.callTool('get_whitelist', {});
    console.log(whitelist.content[0].text);

    // Execute a safe command
    console.log('\nExecuting a safe command (ls -la):');
    const lsResult = await client.callTool('execute_command', {
      command: 'ls',
      args: ['-la'],
    });
    console.log(lsResult.content[0].text);

    // Add a command to the whitelist
    console.log('\nAdding a command to the whitelist:');
    const addResult = await client.callTool('add_to_whitelist', {
      command: 'node',
      securityLevel: 'safe',
      description: 'Execute Node.js scripts',
    });
    console.log(addResult.content[0].text);

    // Try executing a command that requires approval
    console.log('\nExecuting a command that requires approval (mkdir test-dir):');
    try {
      const mkdirResult = await client.callTool('execute_command', {
        command: 'mkdir',
        args: ['test-dir'],
      });
      console.log(mkdirResult.content[0].text);
    } catch (error) {
      console.log('Command requires approval. Getting pending commands...');
      
      // Get pending commands
      const pendingCommands = await client.callTool('get_pending_commands', {});
      const pendingCommandsObj = JSON.parse(pendingCommands.content[0].text);
      
      if (pendingCommandsObj.length > 0) {
        const commandId = pendingCommandsObj[0].id;
        console.log(`Approving command with ID: ${commandId}`);
        
        // Approve the command
        const approveResult = await client.callTool('approve_command', {
          commandId,
        });
        console.log(approveResult.content[0].text);
      }
    }

    // Clean up - remove the test directory
    console.log('\nCleaning up:');
    try {
      // This will fail because 'rm' is forbidden
      const rmResult = await client.callTool('execute_command', {
        command: 'rm',
        args: ['-rf', 'test-dir'],
      });
      console.log(rmResult.content[0].text);
    } catch (error) {
      console.log(`Clean-up failed: ${error.message}`);
      console.log('Note: This is expected because "rm" is a forbidden command');
    }

  } catch (error) {
    console.error('Error:', error);
  } finally {
    // Disconnect from the server
    await client.disconnect();
    console.log('\nDisconnected from Mac Shell MCP Server');
  }
}

main().catch(console.error);

File: examples/client_example.py
import asyncio
import json

from mcp.client.stdio import StdioServerParameters, stdio_client
from mcp.client.session import ClientSession


async def main():
    # Start the server as a child process via stdio
    params = StdioServerParameters(command="python", args=["-m", "your_pkg"])

    async with stdio_client(params) as (read, write):
        session = ClientSession()

        # Initialize session (handshake)
        await session.initialize(read, write)

        # --- List available tools ---
        tools = await session.list_tools()
        print("Available tools:")
        for t in tools:
            print(f"- {t.name}: {t.description}")

        # --- Get the whitelist ---
        print("\nWhitelisted commands:")
        wl_res = await session.call_tool("get_whitelist", {})
        print(wl_res.content[0].text)

        # --- Execute a safe command ---
        print("\nExecuting a safe command (echo 'hello from client'):")
        safe_res = await session.call_tool(
            "execute_command",
            {"command": "echo", "args": ["hello from client"]},
        )
        for block in safe_res.content:
            if getattr(block, "text", ""):
                print(block.text)

        # --- Add a command and require approval (mkdir) ---
        print("\nMarking 'mkdir' as requires_approval and trying it:")
        await session.call_tool(
            "add_to_whitelist",
            {"command": "mkdir", "securityLevel": "requires_approval", "description": "Create dirs"},
        )
        try:
            mk_res = await session.call_tool(
                "execute_command",
                {"command": "mkdir", "args": ["test-dir"]},
            )
            # If the server returned text directly (unlikely in approval path), print it
            for block in mk_res.content:
                if getattr(block, "text", ""):
                    print(block.text)
        except Exception as e:
            print(f"(Expected) Command requires approval: {e}")

        # --- Get pending commands and approve the first one ---
        print("\nGetting pending commands:")
        pend = await session.call_tool("get_pending_commands", {})
        pend_list = json.loads(pend.content[0].text)
        print(json.dumps(pend_list, indent=2))

        if pend_list:
            cmd_id = pend_list[0]["id"]
            print(f"Approving command with ID: {cmd_id}")
            approve_res = await session.call_tool("approve_command", {"commandId": cmd_id})
            for block in approve_res.content:
                if getattr(block, "text", ""):
                    print(block.text)

        # --- Try a forbidden command (rm) to show failure path ---
        print("\nTrying a forbidden command (rm -rf test-dir):")
        try:
            rm_res = await session.call_tool(
                "execute_command",
                {"command": "rm", "args": ["-rf", "test-dir"]},
            )
            for block in rm_res.content:
                if getattr(block, "text", ""):
                    print(block.text)
        except Exception as e:
            print(f"(Expected) Forbidden or error: {e}")

        # Graceful shutdown
        await session.shutdown()


if __name__ == "__main__":
    asyncio.run(main())
